examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,698 @@
+# to preserve to some level. In the following we try to explore smooth,
+# non-monotonic encoding that locally preserves the relative ordering of time
+# features.
+#
+# Trigonometric features
+# ----------------------
+#
+# As a first attempt, we can try to encode each of those periodic features
+# using a sine and cosine transform with the matching period.
+#
+# Each ordinal time feature is transformed into 2 features that together encode
+# equivalent information in a non-monotonic way, and more importantly without
+# any jump between the first and the last value of the periodic range.
+from sklearn.preprocessing import FunctionTransformer
+
+
+def sin_transformer(period):
+    return FunctionTransformer(lambda x: np.sin(x / period * 2 * np.pi))
+
+
+def cos_transformer(period):
+    return FunctionTransformer(lambda x: np.cos(x / period * 2 * np.pi))
+
+
+# %%
+#
+# Let us visualize the effect of this feature expansion on some synthetic hour
+# data with a bit of extrapolation beyond hour=23:
+import pandas as pd
+
+hour_df = pd.DataFrame(
+    np.linspace(0, 26, 1000).reshape(-1, 1),
+    columns=["hour"],
+)
+hour_df["hour_sin"] = sin_transformer(24).fit_transform(hour_df)["hour"]
+hour_df["hour_cos"] = cos_transformer(24).fit_transform(hour_df)["hour"]
+hour_df.plot(x="hour")
+_ = plt.title("Trigonometric encoding for the 'hour' feature")
+
+# %%
+#
+# We can now build a feature extraction pipeline using this strategy:
+cyclic_cossin_transformer = ColumnTransformer(
+    transformers=[
+        ("categorical", one_hot_encoder, categorical_columns),
+        ("month_sin", sin_transformer(12), ["month"]),
+        ("month_cos", cos_transformer(12), ["month"]),
+        ("weekday_sin", sin_transformer(7), ["weekday"]),
+        ("weekday_cos", cos_transformer(7), ["weekday"]),
+        ("hour_sin", sin_transformer(24), ["hour"]),
+        ("hour_cos", cos_transformer(24), ["hour"]),
+    ],
+    remainder="passthrough",
+)
+cyclic_cossin_linear_pipeline = make_pipeline(
+    cyclic_cossin_transformer,
+    RidgeCV(alphas=alphas),
+)
+evaluate(cyclic_cossin_linear_pipeline, X, y, cv=ts_cv)
+
+
+# %%
+#
+# The performance of our linear regression model with this simple feature
+# engineering is a bit better than using the original ordinal time features but
+# worse than using the one-hot encoded time features. We will further analyze
+# possible reasons for this disappointing outcome at the end of this notebook.
+#
+# Periodic spline features
+# ------------------------
+#
+# We can try an alternative encoding of the periodic time-related features
+# using spline transformations with a large enough number of splines, and as a
+# result a larger number of expanded features:
+from sklearn.preprocessing import SplineTransformer
+
+
+def periodic_spline_transformer(period, n_splines=None, degree=3):
+    if n_splines is None:
+        n_splines = period
+    n_knots = n_splines + degree - 1
```

Comment:
I think the formula is `n_knots = n_splines + 1` in the periodic case.

Reply:
With `include_bias=False`, it is even `n_knots = n_splines + 2`, if my math works out:stuck_out_tongue_winking_eye:

Reply:
This is what I found empirically. The docstring could be improved ;)

Reply:
Well, I don't fully understand how splines are defined, but if you consider `SplineTransformer` to be a smoothed version of `KBinsDiscretizer`, then its number of "bins" is `n_knots - 1` in the periodic case, with or without `include_bias`. `include_bias` is only removing one redundant spline (redundant since a sample not in the first n-1 bins is necessarily in the last, assuming the bins cover all cases), but `include_bias` does not change the number of bins.

In the end, the number of splines is arbitrary, but if we want it to make sense that we are using 7 bins for the weekday feature, then we need 8 knots and 6 splines. So I would actually use `n_knots = n_splines + 1`.

(By the way, we might want to include a parameter `include_bias` in `KBinsDiscretizer` with the same mechanism.)

Reply:
If you exclude the intercept, `n` bins correspond to `n - 1` features. As you mention @TomDLT, this is comparable to splines, where `n + 1` knots correspond to `n` bins and thus `n - 1` features in the periodic case.

Using periodic splines with `n + 1` knots for a feature which takes `n` unique values modulo periodicity (e.g. 8 knots for day of the week) is a bad idea though, as these features, when used in a linear model, are mathematically equivalent to binning (which is simpler). 

The benefit of periodic splines over normal splines is that you can encode that days 1 and 7 of the week are as similar as e.g. say 2 and 3. If we use 8 knots (which is equivalent to binning), we lose the information about proximity entirely (as when binning).